spokestack/asr/spokestack/keyword_recognizer.py

```diff
@@ -0,0 +1,195 @@
+        # with the previous sample then cache the previous sample
+        # to be use in the next iteration
+        prev_sample = frame[-1]
+        frame -= self.pre_emphasis * np.append(self._prev_sample, frame[:-1])
+        self._prev_sample = prev_sample
+
+        # fill the sample window to analyze speech containing samples
+        # after each window fill the buffer advances by the hop length
+        # to produce an overlapping window
+        for sample in frame:
+            self.sample_window.write(sample)
+            if self.sample_window.is_full:
+                if context.is_active:
+                    self._analyze(context)
+                self.sample_window.rewind().seek(self.hop_length)
+
+    def _analyze(self, context: SpeechContext) -> None:
+        # read the full contents of the sample window to calculate a single frame
+        # of the STFT by applying the DFT to a real-valued input and
+        # taking the magnitude of the complex DFT
+        frame = self.sample_window.read_all()
+        frame = np.fft.rfft(frame * self._fft_window, n=self._window_size)
+        frame = np.abs(frame).astype(np.float32)
+
+        # compute mel spectrogram
+        self._filter(context, frame)
+
+    def _filter(self, context: SpeechContext, frame) -> None:
+        # add the batch dimension and compute the mel spectrogram with filter model
+        frame = np.expand_dims(frame, 0)
+        frame = self.filter_model(frame)[0]
+
+        # advance the window by 1 and write mel frame to the frame buffer
+        self.frame_window.rewind().seek(1)
+        self.frame_window.write(frame)
+
+        # encode the mel spectrogram
+        self._encode(context)
+
+    def _encode(self, context: SpeechContext) -> None:
+        # read the full contents of the frame window and add the batch dimension
+        # run the encoder and save the output state for autoregression
+        frame = self.frame_window.read_all()
+        frame = np.expand_dims(frame, 0)
+        frame, self.state = self.encode_model(frame, self.state)
+
+        # accumulate encoded samples until size of detection window
+        self.encode_window.rewind().seek(1)
+        self.encode_window.write(frame)
+
+    def _detect(self, context: SpeechContext) -> None:
+        # read the full contents of the encode window and add the batch dimension
+        # calculate a scalar likelihood that the frame contains a keyword
+        # with the detect model
+        frame = self.encode_window.read_all()
+        frame = np.expand_dims(frame, 0)
+        posterior = self.detect_model(frame)[0][0]
+        class_index = np.argmax(posterior, -1)
+        confidence = posterior[class_index]
+        print(confidence)
+
+        if confidence >= self._posterior_threshold:
+            context.transcript = self.classes[class_index]
+            context.confidence = confidence
+            context.event("recognize")
+            _LOG.debug("recognize event")
```

Comment:
since you're already raising events for these, you probably don't need debug logs that say the same thing, especially since you already have a logging statement above